kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/ProducerStateManager.java

```diff
@@ -14,98 +14,175 @@
+                });
+
+        return result;
+    }
+
+    public CompletableFuture<ProducerStateManagerSnapshot> takeSnapshot(Executor executor) {
+        CompletableFuture<ProducerStateManagerSnapshot> result = new CompletableFuture<>();
+        executor.execute(new SafeRunnable() {
+            @Override
+            public void safeRun() {
+                if (mapEndOffset == -1) {
+                    result.complete(null);
+                    return;
+                }
+                ProducerStateManagerSnapshot snapshot = getProducerStateManagerSnapshot();
+                log.info("Taking snapshot for {} at {}", topicPartition, snapshot);
+                producerStateManagerSnapshotBuffer
+                        .write(snapshot)
+                        .whenComplete((res, error) -> {
+                            if (error != null) {
+                                result.completeExceptionally(error);
+                            } else {
+                                log.info("Snapshot for {} taken at offset {}",
+                                        topicPartition, snapshot.getOffset());
+                                result.complete(snapshot);
+                            }
+                        });
+            }
+        });
+        return result;
+    }
+
+    void maybeTakeSnapshot(Executor executor) {
+        if (mapEndOffset == -1) {
+            return;
+        }
+        long now = System.currentTimeMillis();
+        long deltaFromLast = (now - lastSnapshotTime) / 1000;
+        if (log.isDebugEnabled()) {
+            log.debug("maybeTakeSnapshot deltaFromLast {} vs kafkaTxnProducerStateTopicSnapshotIntervalSeconds {} ",
+                    deltaFromLast, kafkaTxnProducerStateTopicSnapshotIntervalSeconds);
+        }
+        if (deltaFromLast < kafkaTxnProducerStateTopicSnapshotIntervalSeconds) {
+            return;
+        }
+        lastSnapshotTime = now;
+
+
+        // take the snapshot in this thread, that is the same thread
+        // that executes mutations
+        ProducerStateManagerSnapshot snapshot = getProducerStateManagerSnapshot();
+
+        // write to Pulsar in another thread, and also ignore errors
+        executor.execute(new SafeRunnable() {
+            @Override
+            public void safeRun() {
+                producerStateManagerSnapshotBuffer
+                        .write(snapshot)
+                        .whenComplete((res, error) -> {
+                            if (error == null) {
+                                log.info("Snapshot for {} taken at offset {} written",
+                                        topicPartition, snapshot.getOffset());
+                            } else {
+                                log.info("Error writing snapshot for {} taken at offset {}",
+                                        topicPartition, snapshot.getOffset(), error);
+                            }
+                        });
+            }
+        });
```

Comment:
Could we use the method `takeSnapshot` instead of this?

Reply:
Yes, updated